src/WignerSymbols.jl

```diff
@@ -242,6 +247,85 @@ function racahW(T::Type{<:Real}, j₁, j₂, J, j₃, J₁₂, J₂₃)
+        (ishalfinteger(jᵢ) && jᵢ >= zero(jᵢ)) || throw(DomainError("invalid jᵢ", jᵢ))
+    end
+    return _wigner9j(T, HalfInteger.((j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉))...)
+end
+
+const _perms9j = [([1, 2, 3], [1, 2, 3]) ([1, 2, 3], [1, 3, 2]) ([1, 2, 3], [2, 1, 3]) ([1, 2, 3], [2, 3, 1]) ([1, 2, 3], [3, 1, 2]) ([1, 2, 3], [3, 2, 1]);
+                  ([1, 3, 2], [1, 2, 3]) ([1, 3, 2], [1, 3, 2]) ([1, 3, 2], [2, 1, 3]) ([1, 3, 2], [2, 3, 1]) ([1, 3, 2], [3, 1, 2]) ([1, 3, 2], [3, 2, 1]);
+                  ([2, 1, 3], [1, 2, 3]) ([2, 1, 3], [1, 3, 2]) ([2, 1, 3], [2, 1, 3]) ([2, 1, 3], [2, 3, 1]) ([2, 1, 3], [3, 1, 2]) ([2, 1, 3], [3, 2, 1]);
+                  ([2, 3, 1], [1, 2, 3]) ([2, 3, 1], [1, 3, 2]) ([2, 3, 1], [2, 1, 3]) ([2, 3, 1], [2, 3, 1]) ([2, 3, 1], [3, 1, 2]) ([2, 3, 1], [3, 2, 1]);
+                  ([3, 1, 2], [1, 2, 3]) ([3, 1, 2], [1, 3, 2]) ([3, 1, 2], [2, 1, 3]) ([3, 1, 2], [2, 3, 1]) ([3, 1, 2], [3, 1, 2]) ([3, 1, 2], [3, 2, 1]);
+                  ([3, 2, 1], [1, 2, 3]) ([3, 2, 1], [1, 3, 2]) ([3, 2, 1], [2, 1, 3]) ([3, 2, 1], [2, 3, 1]) ([3, 2, 1], [3, 1, 2]) ([3, 2, 1], [3, 2, 1])]
+
+const _signs9j = [1 -1 -1 1 1 -1;
+                  -1 1 1 -1 -1 1;
+                  -1 1 1 -1 -1 1;
+                  1 -1 -1 1 1 -1;
+                  1 -1 -1 1 1 -1;
+                  -1 1 1 -1 -1 1]
+
+function _wigner9j(T::Type{<:Real}, j₁::HalfInteger, j₂::HalfInteger, j₃::HalfInteger, 
+                                    j₄::HalfInteger, j₅::HalfInteger, j₆::HalfInteger,
+                                    j₇::HalfInteger, j₈::HalfInteger, j₉::HalfInteger)
+  
+    # check triangle conditions 
+    if !(δ(j₁, j₂, j₃) && δ(j₄, j₅, j₆) && δ(j₇, j₈, j₉) && δ(j₁, j₄, j₇) && δ(j₂, j₅, j₈) && δ(j₃, j₆, j₉))
+        return zero(T)
+    end
+   
+    # dictionary lookup, check all 72 permutations 
+    k = [j₁ j₂ j₃; j₄ j₅ j₆; j₇ j₈ j₉]
+    for (p, m) in zip(_perms9j, _signs9j)
+        kk = Tuple(reshape(k[p...], 9))
+        kkT = Tuple(reshape(transpose(k[p...]), 9))
+        if haskey(Wigner9j, kk)
+            r, s = Wigner9j[kk]
+            return m * _convert(T, s) * convert(T, signedroot(r))
+        elseif haskey(Wigner9j, kkT)
+            r, s = Wigner9j[kkT]
+            return m * _convert(T, s) * convert(T, signedroot(r)) 
+        end
+    end
+
+    # order irrelevant: product remains the same
+    n₁, d₁ = Δ²(j₁, j₂, j₃)
+    n₂, d₂ = Δ²(j₄, j₅, j₆)
+    n₃, d₃ = Δ²(j₇, j₈, j₉)
+    n₄, d₄ = Δ²(j₁, j₄, j₇)
+    n₅, d₅ = Δ²(j₂, j₅, j₈)
+    n₆, d₆ = Δ²(j₃, j₆, j₉) 
+
+    snum, rnum = splitsquare(n₁ * n₂ * n₃ * n₄ * n₅ * n₆)
+    sden, rden = splitsquare(d₁ * d₂ * d₃ * d₄ * d₅ * d₆)
+    snum, sden = divgcd!(snum, sden)
+    rnum, rden = divgcd!(rnum, rden) 
+    s = Base.unsafe_rational(convert(BigInt, snum), convert(BigInt, sden))
+    r = Base.unsafe_rational(convert(BigInt, rnum), convert(BigInt, rden))
+    s *= compute9jseries(j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉)
```

Comment:
So if I understand correctly, the result of `compute9jseries` here is a `BigInt`; there are no denominators involved. Maybe then it could be slightly faster still to first take out primefactors from `sden` and only then construct the rational, i.e. something like

```julia
snum2 = compute9jseries(j₁, j₂, j₃, j₄, j₅, j₆, j₇, j₈, j₉)
for n = 1:length(sden.powers)
    p = bigprime(n)
    while sden.powers[n] > 0
        q, r = divrem(snum2, p)
        if iszero(r)
            snum2 = q
            sden.powers[n] -= 1
        else
            break
        end
    end
end
Base.unsafe_rational(convert(BigInt, snum)*snum2, convert(BigInt, sden))
```

Reply:
Yeah - the series factor in the [9j formula given by Wei](https://pubs.aip.org/aip/cip/article/12/6/632/510234/New-formula-for-9-j-symbols-and-their-direct) is always just an integer so `compute9jseries` returns a `BigInt` (actually I think it returns a `BigHalfInt`, but it should actually always be integer so I'll add that conversion to the end of `compute9jseries` for posterity). So that's a good catch - I'll add that optimization.